### PR TITLE
[ENH] New analytic integral property for models

### PR DESCRIPTION
**Pull request by pllim**
This is a fully functional proof of concept for #5033.

Added new `integral` property to `modeling` core. Implemented analytic integral for `Box1D` and `Box2D`. Added relevant tests.

TODO:
- [x] Remove "ugly hack" for `Box2D` integral? Yes.
- [ ] Wait for Nadia's review.
- [ ] Wrap up review discussions.
- [ ] Add test for new ramp model by defining it in `modeling/tests/example_models.py`.
- [ ] Add change log entry.

c/c @mcara, @mhvk

### CODE REVIEW

**Comment by PaulKuin** [comment]
The modeling Gaussian1d integral is useful for modeling line profiles. The currently defined function integrates to A\sigma\sqrt(2\pi). It should be easy to add a method for this.

**Comment by pllim** [comment]
I would prefer to get more feedback to the underlying machinery before adding more integrals. But a Gaussian would be nice, thanks.

**Comment by mhvk** [comment]
This reminds me I should be reviewing this...  Do ping me if I still haven't done so by the end of the week...

**Comment by pllim** [comment]
I should c/c @nden and @embray too, in case they also want to review...

**Comment by pllim** [comment]
@mhvk -- ping...

File: astropy/modeling/functional_models.py [hunk 1]
```diff
@@ -1315,6 +1315,75 @@ def __init__(self):
         raise ModelDefinitionError("Not implemented")
 
 
+class RampForBox1D(Fittable1DModel):
```
Comment (mhvk) [inline astropy/modeling/functional_models.py]: Should we just call it `Ramp1D`? This does mean initializing it with an amplitude that is box amplitude / box width, but I think it means also that the function is more generally useful.

Although possibly even better would be to combine a new `Clip` model with `Linear1D`...

File: astropy/modeling/functional_models.py [hunk 2]
```diff
@@ -1469,6 +1551,53 @@ def bounding_box(self):
         return ((self.y_0 - dy, self.y_0 + dy),
                 (self.x_0 - dx, self.x_0 + dx))
 
+    def integral(self, axis='x'):
```
Comment (mhvk) [inline astropy/modeling/functional_models.py]: Don't know that I like the 2D implementation, as it implicitly assumes the other coordinate equals 0.

**Comment by mhvk** [comment]
@pllim - finally had at least a bit of a look. In general, the 1D case looks good (except that I'd hope to avoid creating very specific integral models; if needed, make a `Ramp1D`, or somehow use the existing `Linear1D` with its bounding box or so).

For 2D, I'm not convinced this is the right approach; if you integrate over one axis, you somehow have to be able to specify the value in the other coordinate. Since I don't think you actually need 2-D yet, maybe we should simply punt on that for now?

@embray, @nden - I'm not familiar enough with the modelling framework to feel that I can really do justice in a review; could you have a look as well?

**Comment by nden** [comment]
@pllim I won't be able to look at this before the end of Oct.ober.

**Comment by mcara** [comment]
I agree with @mhvk that we should hold back the 2D case. For 1D case situation is much simple and easily doable.

### 1D Case

I would suggest that 1D models have a 'primitive' method that returns the primitive (antiderivative) function when available or ``None`` when primitive is not implemented. The 'integral' method then would return definite integral (=flux?) which is easily computed using 'primitive' when available or switch to numerical integration.

### 2D Case

This is much more "difficult" because much depends on the domain of integration. That is, a "primitive" in this case, generally speaking, will not be able to provide the "flux" in a simple manner such as ``flux=F(x2,y2)-F(x1,y1)``. Something like a 2D primitive would work like this only when the domain of integration is rectangular and ``x1=0`` and/or ``y2=0`` (and if it is "or" then the primitive will have to be implemented assuming that one of the coordinates x1 or y1 is zero and it will not work if the *other* coordinate is zero but the first one is not). In order to implement analytical 2D integration I think first the problem of specifying the domain of integration will need to be solved. Even then, it is not straightforward how to implement a primitive.

What could be possible to do right now for the 2D case is to implement analytical integration for circular domains because in that case (=when using complete circles as opposite to sectors), the problem can be reduced to 1D integration along radial direction.

==========

So, I think it would be wiser to postpone any 2D implementation until better concepts could be developed/agreed upon.

**Comment by pllim** [comment]
@mcara, feel free to open an alternate PR with your suggestion.

**Comment by mhvk** [comment]
@mcara - I think that is a good summary of the issues, and the idea to separate out the definite from the indefinite integral makes sense (and I like the suggested names). @pllim - wouldn't this get you closer to what you originally hoped to do, which is to have the ability to calculate definite integrals?

**Comment by pllim** [comment]
Maybe? I am not sure what "primitive" means in terms of implementation and I am busy with other stuff probably for the next few weeks, so faster if Mihai submits a PR with his idea...

File: astropy/modeling/core.py [hunk 3]
```diff
@@ -265,6 +266,76 @@ def _create_inverse_property(cls, members):
...
+        Takes the function's Signature as an argument since that is already
+        computed in _create_integral_property, so no need to duplicate that
+        effort.
+        """
+
+        def __call__(self, **kwargs):
+            return func(self._model, **kwargs)
+
+        kwargs = []
+        for idx, param in enumerate(sig.parameters.values()):
```
Comment (pllim) [inline astropy/modeling/core.py]: Indeed. I copied this from `_create_bounding_box_subclass()`, so I'll change this over there as well.

File: astropy/modeling/core.py [hunk 4]
```diff
@@ -906,6 +980,86 @@ def has_user_inverse(self):
...
+            # subclass, so we call it with its default arguments and return an
+            # instance of it (that can be called to recompute the integral
+            # with any optional parameters)
+            # (In other words, in this case self._integral is a *class*)
+            integral = self._integral(_model=self)()
+            return self._integral(_model=self)
+
+    @integral.setter
+    def integral(self, integral):
+        if integral is None:
```
Comment (pllim) [inline astropy/modeling/core.py]: Yeah, I don't understand why I did that either. Good catch. Will fix.

File: astropy/modeling/functional_models.py (hunk 1 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: The current formula is very "boxy" (i.e., uses `A`, `w`, and `x_0`). How do you envision usage of a generalized `Ramp1D`? For instance, if I want to use `Ramp1D` to integrate `Trapezoid1D`, the current input parameters will not be sufficient.

In regards to `Clip` model, I would rather implement that in a separate PR, lest this one gets too big, and hence hard to review. This one should be focused on the underlying machinery and a working proof of concept (for `Box1D`).

File: astropy/modeling/functional_models.py (hunk 2 above)
Comment (pllim) [inline astropy/modeling/functional_models.py]: Looks like consensus is to remove 2D implementation, so this won't be an issue when that is removed.

**Comment by mhvk** [comment]
@pllim - `primitive` would be the indefinite integral, so for Box1D it would return Ramp1D; `integrate` would then be the definite integral, i.e., you'd pass in start and stop, and get out a number (as I think you had originally). The idea would be that when there is a `primitive` defined, `integrate` could just use that, as it would be very fast, but if there is no `primitive`, at least in principle `integrate` could use a numerical method to do the integration (which of course does not have to come in this PR!).

p.s.  Am travelling so a bit pressed for time, but I will try to have a look again once you let me know I should.
p.s.2 I don't see that a `Ramp1D` model could ever be used for a trapezoid: that needs parabolas for the slanted edges.

**Comment by pllim** [comment]
@mhvk , your suggestion to replace `enumerate(...)` with `[1:]` produced `TypeError: 'odict_values' object is not subscriptable` exception, so I will have to revert that change. FYI.

File: astropy/modeling/core.py [hunk 5]
```diff
@@ -266,6 +267,73 @@ def _create_inverse_property(cls, members):
...
+        computed in _create_integral_property, so no need to duplicate that
+        effort.
+        """
+
+        def __call__(self, **kwargs):
+            return func(self._model, **kwargs)
+
+        # Already checked for sig.parameters > 1 in _create_integral_property()
+        kwargs = []
+        for param in sig.parameters.values()[1:]:  # [0] = self
```
Comment (pllim) [inline astropy/modeling/core.py]: Huh... `islice` is already imported anyway!

**Comment by pllim** [comment]
@mhvk 

Re: primitive. Okay, I *think* I get it now. But I disagree on falling back on numerical integration automatically by default. For some models (e.g., `Shift` and `Scale`), integration simply does not make sense.

**Comment by pllim** [comment]
@mhvk and @mcara , I think I addressed all your comments so far. Please have another look (if nothing else, at the tests for a feel on the API). If you don't mind enabling `integrate()` for everything including models that don't really need it (like `Shift` and `Scale`), one idea is to fall back to trapezoid integration using [np.trapz](https://docs.scipy.org/doc/numpy/reference/generated/numpy.trapz.html) in the absence of `model.primitive`.

**Comment by mhvk** [comment]
@pllim - this looks all OK.  And is a great addition!

A few more things:
1. While the travis failure seems a red herring (restarted the failed job), it does seem you do not yet cover part of the primitive creation in model and util; if you can think of some good examples, it would be nice to have tests that add that coverage.
2. This will need to get documented. Were you planning to do that here, now that the scheme is set? Or perhaps as part of a bit larger effort to add more primitives? If the latter, maybe raise an issue to remind us of that.
3. We do need a changelog entry.

**Comment by pllim** [comment]
@mcara communicated to me that he now prefers `primitive` to always be a method and only returns a *function*, not model. I am hesitating to continue work on this PR until a general consensus could be reach on how you want things done here. For my own usage, all I wanted is an integrated result (flux).

**Comment by mcara** [comment]
Look, I am not a major "stakeholder" in this, so treat my opinions as such and in the end do what you think is the best for astropy going forward. Below I will comment on several general issues mentioned previously in this thread.

===== On the issue that ``primitive`` return Models

I know I am the one who proposed (=wondered) that ``primitive`` return Models. I think this idea may work very well for the 1D case but I have trouble imagining this working for the 2D case. On the other hand, I also cannot see how a ``primitive`` would work for 2D case at all. So, if we are ready to give up on 2D case or if we accept that 2D case will be handled differently from the 1D case, then I think it is OK for ``primitive`` to return a new Model. Otherwise, I would [probably] stick with returning functions.

Ask yourself the following question: is it worth introducing new models into ``astropy.modeling`` to allow for integrated models?

On the other hand, are we going to support ``integral`` and/or ``primitive`` on **compound** models? If so, returning models may be more useful since each model could contain ``derivative`` property that returns derivative models (so we will need to add those new models to the modeling framework as well) and these derivatives together with primitives of the individual models [within a compound model] can be used to derive the primitives of the compound models. **However**, these new primitives (of arbitrary compound models) may not be part of modeling framework! How are we going to deal with this???

Maybe the answer would be to use https://github.com/sympy ??? Whatever the answer is, I think this issue deserves a thorough investigation and thinking as it will shape the way integration is done in modeling for 1D and possibly 2D models.

===== On the issue of integral not making sense for some models

I completely agree with you that integration makes little sense for many of the ``astropy.modeling`` models. For this reason, after studying [inheritance diagram](http://docs.astropy.org/en/stable/modeling/#class-inheritance-diagram) for model classes, it seems to me that integration only makes sense for the following categories/classes:

- ``Fittable1DModel``
- Some ``PolynomialModel`` subclasses (exclude 2D cases for now): ``Chebyshev1D``, ``Legendre1D``, and ``Polynomial1D``.
- ``SIP`` and ``InverseSIP``.

For this reason I am surprised that you are trying to add ``integral`` and ``primitive`` properties to the very base ``Model``.

===== On ``_create_primitive_property`` fanciness

Present implementation seems to me too complicated for what needs to be achieved will all these fancy ``_create_primitive_property`` things but I have to admit I may not understand the innate beauty of astropy modeling machinery.

**Comment by pllim** [comment]
My use case is only integrating flux for 1D models, so I have no motivation to think too much about integration for 2D models. If you don't want to even enable it for 2D, I can move the base stuff further down so only 1D models get that capability. I'll wait for @nden 's review before proceeding.

**Comment by pllim** [comment]
This received many conflicting suggestions and appears to be going nowhere, so closing. I'll just keep it in my "synphot" package for now. If there is interest, we can revisit after models support units.

**Comment by mhvk** [comment]
I know it is your PR, but I think this was more than nice enough to keep open, so that someone else can work on it!  So, can I suggest you reconsider? (Since you deleted the branch, I cannot reopen...)

**Comment by pllim** [comment]
@mhvk , the original request is at #5033 (still open). 😄

**Pull request closed without merging**